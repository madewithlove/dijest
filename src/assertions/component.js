import renderer from 'react-test-renderer';

/**
 * Assert that a component equals X
 */
export default actual => {
    if (typeof actual.toJSON === 'undefined') {
        actual = renderer.create(actual);
    }

    const tree = actual.toJSON();
    expect(tree).toMatchSnapshot();

    return {actual, tree};
};
