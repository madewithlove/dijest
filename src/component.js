import {mount, render} from 'enzyme';
import {matchesSnapshot} from './helpers/matchesSnapshot';
import toEnzymeWrapper from './helpers/toEnzymeWrapper';

export default function component(description, tested, callback) {
    if (!tested) {
        return matchesSnapshot(toEnzymeWrapper(description));
    }

    it(description, () => {
        const wrapper = toEnzymeWrapper(tested, callback ? mount : render);
        if (callback) {
            callback(wrapper, matchesSnapshot);
        } else {
            matchesSnapshot(wrapper);
        }
    });
}
