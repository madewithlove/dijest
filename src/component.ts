import { mount, render } from 'enzyme';
import matchesSnapshot from './helpers/matchesSnapshot';
import toEnzymeWrapper, { EnzymeWrapper } from './helpers/toEnzymeWrapper';
import { ComponentClass, ReactElement, StatelessComponent } from 'react';
import withJestOptions, { JestOptions } from './helpers/withJestOptions';
import ComponentTester from './Testers/ComponentTester';

export type ComponentFactory =
    | ReactElement<any>
    | StatelessComponent
    | ComponentClass;

export type ComponentCallback = (
    wrapper: EnzymeWrapper,
    matchesSnapshot: Function,
) => void;

function component(
    description: string | ComponentFactory,
    tested?: ComponentFactory,
    callback?: ComponentCallback | string,
    it?: jest.It,
) {
    const tester = new ComponentTester().setTester(it);
    const numberArgs = [description, tested, callback].filter(
        value => typeof value !== 'undefined',
    ).length;

    switch (true) {
        case numberArgs === 1 && typeof description !== 'string':
            return tester.setComponent(description as ComponentFactory).run();

        case numberArgs === 2:
            return tester
                .setDescription(description as string)
                .setComponent(tested)
                .run();

        case numberArgs === 3 && typeof callback === 'string':
            return tester
                .setDescription(description as string)
                .setComponent(tested)
                .onlySelector(callback as string)
                .run();

        case numberArgs === 3 && typeof callback === 'function':
            return tester
                .setDescription(description as string)
                .setComponent(tested)
                .setCallback(callback as Function)
                .run();
    }
}

export default withJestOptions(component);
