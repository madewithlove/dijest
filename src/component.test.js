import React from 'react';
import component from './component';

const DummyComponent = ({text}) => <div><p>{text || 'foobar'}</p></div>;

component('can write snapshot test', <DummyComponent text="hello" />);
component('can pass component as function', DummyComponent);
component('can receive wrapper', DummyComponent, (wrapper, snapshot) => {
   snapshot(wrapper.find('p'));
});
