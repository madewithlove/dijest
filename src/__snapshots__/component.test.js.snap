exports[`test can pass component as function 1`] = `
<div>
  <p>
    foobar
  </p>
</div>
`;

exports[`test can receive wrapper 1`] = `
<p>
  foobar
</p>
`;

exports[`test can write snapshot test 1`] = `
<div>
  <p>
    hello
  </p>
</div>
`;
